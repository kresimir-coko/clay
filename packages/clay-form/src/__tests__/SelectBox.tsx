/**
 * SPDX-FileCopyrightText: Â© 2020 Liferay, Inc. <https://liferay.com>
 * SPDX-License-Identifier: BSD-3-Clause
 */

import {ClaySelectBox} from '..';
import {cleanup, fireEvent, render} from '@testing-library/react';
import React from 'react';

const options = [
	{
		label: 'Option 1',
		value: '1',
	},
	{
		label: 'Option 2',
		value: '2',
	},
	{
		label: 'Option 3',
		value: '3',
	},
];

describe('Rendering', () => {
	afterEach(cleanup);

	it('renders ClaySelectBox', () => {
		const {container} = render(
			<ClaySelectBox
				aria-label="Select Box Label"
				id="mySelectBoxId"
				items={options}
				onChange={() => {}}
				value={'1'}
			/>
		);

		expect(container).toMatchSnapshot();
	});

	it('renders ClaySelectBox with multiple options selected', () => {
		const {container} = render(
			<ClaySelectBox
				aria-label="Select Box Label"
				id="mySelectBoxId"
				items={options}
				multiple
				onChange={() => {}}
				value={['1', '2']}
			/>
		);

		expect(container).toMatchSnapshot();
	});

	it('renders ClaySelectBox with buttons to reorder options', () => {
		const {container} = render(
			<ClaySelectBox
				aria-label="Select Box Label"
				id="mySelectBoxId"
				items={options}
				multiple
				onChange={() => {}}
				onItemsChange={() => {}}
				showArrows
				spritemap="/path/to/some/resource.svg"
				value={['1']}
			/>
		);

		expect(container).toMatchSnapshot();
	});
});

describe('Interactions', () => {
	const handleOnItemsChange = jest.fn();

	afterEach(() => {
		handleOnItemsChange.mockReset();

		cleanup();
	});

	it('changes order of options in ClaySelectBox when reorder up button is clicked', () => {
		const {container} = render(
			<ClaySelectBox
				aria-label="Select Box Label"
				id="mySelectBoxId"
				items={options}
				multiple
				onChange={() => {}}
				onItemsChange={handleOnItemsChange}
				showArrows
				spritemap="/path/to/some/resource.svg"
				value={['2']}
			/>
		);

		const reorderUpButton = container.querySelector(
			'.select-box-order-button-up'
		);

		fireEvent.click(reorderUpButton as HTMLButtonElement, {});

		expect(handleOnItemsChange).toHaveBeenCalled();

		expect(handleOnItemsChange.mock.calls[0][0]).toEqual([
			{label: 'Option 2', value: '2'},
			{label: 'Option 1', value: '1'},
			{label: 'Option 3', value: '3'},
		]);
	});

	it('selects multiple options', () => {
		const {container} = render(
			<ClaySelectBox
				aria-label="Select Box Label"
				id="mySelectBoxId"
				items={options}
				multiple
				onChange={() => {}}
				onItemsChange={() => {}}
				showArrows
				spritemap="/path/to/some/resource.svg"
				value={[]}
			/>
		);

		const optionElement = container.querySelector('option');

		fireEvent.change(optionElement as HTMLOptionElement, {
			target: {value: ['1', '2']},
		});

		expect(container.querySelector('select')!.value).toEqual('1,2');
	});
});
