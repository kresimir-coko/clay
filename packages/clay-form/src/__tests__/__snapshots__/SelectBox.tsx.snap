// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering renders ClaySelectBox 1`] = `
<div>
  <div
    class="form-group-autofit select-box"
  >
    <div
      class="form-group-item"
    >
      <select
        class="form-control select-box-select"
      >
        <option
          class="select-box-option"
          value="1"
        >
          Option 1
        </option>
        <option
          class="select-box-option"
          value="2"
        >
          Option 2
        </option>
        <option
          class="select-box-option"
          value="3"
        >
          Option 3
        </option>
      </select>
    </div>
  </div>
</div>
`;

exports[`Rendering renders ClaySelectBox with buttons to reorder options 1`] = `
<div>
  <div
    class="form-group-autofit select-box"
  >
    <div
      class="form-group-item"
    >
      <select
        class="form-control select-box-select"
        multiple=""
      >
        <option
          class="select-box-option"
          value="1"
        >
          Option 1
        </option>
        <option
          class="select-box-option"
          value="2"
        >
          Option 2
        </option>
        <option
          class="select-box-option"
          value="3"
        >
          Option 3
        </option>
      </select>
      <div
        class="select-box-order-buttons btn-group"
        role="group"
      >
        <button
          class="select-box-order-button select-box-order-button-up btn btn-monospaced btn-secondary"
          type="button"
        >
          <svg
            class="lexicon-icon lexicon-icon-caret-top"
            role="presentation"
          >
            <use
              xlink:href="/path/to/some/resource.svg#caret-top"
            />
          </svg>
        </button>
        <button
          class="select-box-order-button select-box-order-button-down btn btn-monospaced btn-secondary"
          type="button"
        >
          <svg
            class="lexicon-icon lexicon-icon-caret-bottom"
            role="presentation"
          >
            <use
              xlink:href="/path/to/some/resource.svg#caret-bottom"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Rendering renders ClaySelectBox with multiple options selected 1`] = `
<div>
  <div
    class="form-group-autofit select-box"
  >
    <div
      class="form-group-item"
    >
      <select
        class="form-control select-box-select"
        multiple=""
      >
        <option
          class="select-box-option"
          value="1"
        >
          Option 1
        </option>
        <option
          class="select-box-option"
          value="2"
        >
          Option 2
        </option>
        <option
          class="select-box-option"
          value="3"
        >
          Option 3
        </option>
      </select>
    </div>
  </div>
</div>
`;
